File: chunk_player.py
"""
Chunk Player for ADK AI Data Protocol

Replays recorded chunks from JSONL files for testing and debugging.

Usage:
    from chunk_player import ChunkPlayer

    # Create player
    player = ChunkPlayer(
        session_dir="./chunk_logs/session-2025-12-14-123456",
        location="backend-adk-event"
    )

    # Replay chunks
    async for chunk_entry in player.play(mode="fast-forward"):
        # Process chunk_entry.chunk
        print(chunk_entry)

Playback Modes:
    real-time: Replay with original timing (based on timestamps)
    fast-forward: Replay as fast as possible (no delays)
    step: Manual step-by-step (requires next() call)
"""

import asyncio
import json
import time
from collections.abc import AsyncGenerator
from pathlib import Path
from typing import Any, Literal

# Import from chunk_logger for consistent types
from chunk_logger import ChunkLogEntry, LogLocation

PlaybackMode = Literal["real-time", "fast-forward", "step"]


class ChunkPlayer:
    """
    Chunk player for replaying recorded data flow.

    Reads chunks from JSONL files and yields them with timing control.
    """

    def __init__(
        self,
        session_dir: str | Path,
        location: LogLocation,
    ):
        """
        Initialize chunk player.

        Args:
            session_dir: Session directory containing JSONL files
            location: Location to replay (e.g., "backend-adk-event")
        """
        self._session_dir = Path(session_dir)
        self._location = location
        self._jsonl_file = self._session_dir / f"{location}.jsonl"

        if not self._jsonl_file.exists():
            msg = f"JSONL file not found: {self._jsonl_file}"
            raise FileNotFoundError(msg)

    async def play(
        self,
        mode: PlaybackMode = "fast-forward",
    ) -> AsyncGenerator[ChunkLogEntry, None]:
        """
        Replay chunks from JSONL file.

        Args:
            mode: Playback mode (real-time/fast-forward/step)

        Yields:
            ChunkLogEntry: Chunk entries in sequence
        """
        entries = self._load_entries()

        if mode == "fast-forward":
            # Yield as fast as possible
            for entry in entries:
                yield entry
        elif mode == "real-time":
            # Yield with original timing
            start_time = time.time()
            first_timestamp = entries[0].timestamp if entries else 0

            for entry in entries:
                # Calculate delay based on original timestamp
                elapsed_ms = (time.time() - start_time) * 1000
                target_ms = entry.timestamp - first_timestamp
                delay_ms = target_ms - elapsed_ms

                if delay_ms > 0:
                    await asyncio.sleep(delay_ms / 1000)

                yield entry
        elif mode == "step":
            # Manual step-by-step (for interactive debugging)
            # Note: Requires external control mechanism
            # For now, just yield with small delay
            for entry in entries:
                yield entry
                await asyncio.sleep(0.1)  # Small delay for step mode
        else:
            msg = f"Invalid playback mode: {mode}"
            raise ValueError(msg)

    def _load_entries(self) -> list[ChunkLogEntry]:
        """
        Load all entries from JSONL file.

        Returns:
            List of ChunkLogEntry sorted by sequence_number
        """
        entries: list[ChunkLogEntry] = []

        with open(self._jsonl_file, encoding="utf-8") as f:
            for line_no, line in enumerate(f, start=1):
                line = line.strip()
                if not line:
                    continue

                try:
                    data = json.loads(line)
                    entry = ChunkLogEntry(
                        timestamp=data["timestamp"],
                        session_id=data["session_id"],
                        mode=data["mode"],
                        location=data["location"],
                        direction=data["direction"],
                        sequence_number=data["sequence_number"],
                        chunk=data["chunk"],
                        metadata=data.get("metadata"),
                    )
                    entries.append(entry)
                except (json.JSONDecodeError, KeyError) as e:
                    msg = f"Invalid JSONL at line {line_no}: {e}"
                    raise ValueError(msg) from e

        # Sort by sequence_number to ensure correct order
        entries.sort(key=lambda e: e.sequence_number)

        return entries

    def get_stats(self) -> dict[str, Any]:
        """
        Get statistics about the recorded session.

        Returns:
            Dictionary with stats (count, duration, etc.)
        """
        entries = self._load_entries()

        if not entries:
            return {
                "count": 0,
                "duration_ms": 0,
                "first_timestamp": None,
                "last_timestamp": None,
            }

        first_timestamp = entries[0].timestamp
        last_timestamp = entries[-1].timestamp
        duration_ms = last_timestamp - first_timestamp

        return {
            "count": len(entries),
            "duration_ms": duration_ms,
            "first_timestamp": first_timestamp,
            "last_timestamp": last_timestamp,
            "location": self._location,
            "session_dir": str(self._session_dir),
        }

File: lib/chunk_logs/chunk-player.ts
/**
 * Chunk Player for ADK AI Data Protocol (Browser)
 *
 * ⚪ ADK-Independent - Works with all modes (adk-bidi, adk-sse, gemini)
 *
 * Replays recorded chunks from JSONL files for testing and debugging.
 *
 * Dependencies:
 * - AI SDK v6 (UIMessageChunk types)
 * - chunk-logger.ts (JSONL format)
 *
 * Supported Modes:
 * - All modes: Replays any recorded chunk sequence
 *
 * Usage:
 *     import { ChunkPlayer } from './chunk-player';
 *
 *     // Create player from JSONL content
 *     const player = new ChunkPlayer(jsonlContent);
 *
 *     // Replay chunks
 *     for await (const entry of player.play({ mode: 'fast-forward' })) {
 *         // Process entry.chunk
 *         console.log(entry);
 *     }
 *
 * Playback Modes:
 *     real-time: Replay with original timing (based on timestamps)
 *     fast-forward: Replay as fast as possible (no delays)
 *     step: Manual step-by-step (requires next() call)
 */

import type { ChunkLogEntry } from "./chunk-logger";

export type PlaybackMode = "real-time" | "fast-forward" | "step";

export interface PlayOptions {
  mode?: PlaybackMode;
}

export interface PlayerStats {
  count: number;
  duration_ms: number;
  first_timestamp: number | null;
  last_timestamp: number | null;
}

/**
 * Chunk player for replaying recorded data flow.
 *
 * Reads chunks from JSONL content and yields them with timing control.
 */
export class ChunkPlayer {
  private entries: ChunkLogEntry[];

  /**
   * Initialize chunk player.
   *
   * @param jsonlContent - JSONL file content as string
   */
  constructor(jsonlContent: string) {
    this.entries = this.parseJsonl(jsonlContent);
  }

  /**
   * Create player from File object (uploaded JSONL file).
   *
   * @param file - JSONL file
   * @returns Promise<ChunkPlayer>
   */
  static async fromFile(file: File): Promise<ChunkPlayer> {
    const content = await file.text();
    return new ChunkPlayer(content);
  }

  /**
   * Create player from URL (fetch JSONL from server).
   *
   * @param url - URL to JSONL file
   * @returns Promise<ChunkPlayer>
   */
  static async fromUrl(url: string): Promise<ChunkPlayer> {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to fetch JSONL: ${response.statusText}`);
    }
    const content = await response.text();
    return new ChunkPlayer(content);
  }

  /**
   * Parse JSONL content into chunk entries.
   *
   * @param jsonlContent - JSONL file content
   * @returns Array of ChunkLogEntry
   */
  private parseJsonl(jsonlContent: string): ChunkLogEntry[] {
    const lines = jsonlContent.split("\n");
    const entries: ChunkLogEntry[] = [];

    for (let lineNo = 0; lineNo < lines.length; lineNo++) {
      const line = lines[lineNo].trim();
      if (!line) continue; // Skip empty lines

      try {
        const entry = JSON.parse(line) as ChunkLogEntry;
        entries.push(entry);
      } catch (error) {
        throw new Error(
          `Invalid JSONL at line ${lineNo + 1}: ${error instanceof Error ? error.message : String(error)}`,
        );
      }
    }

    // Sort by sequence_number to ensure correct order
    entries.sort((a, b) => a.sequence_number - b.sequence_number);

    return entries;
  }

  /**
   * Replay chunks as async generator.
   *
   * @param options - Playback options
   * @yields ChunkLogEntry entries in sequence
   */
  async *play(options: PlayOptions = {}): AsyncGenerator<ChunkLogEntry> {
    const { mode = "fast-forward" } = options;

    if (mode === "fast-forward") {
      // Yield as fast as possible
      for (const entry of this.entries) {
        yield entry;
      }
    } else if (mode === "real-time") {
      // Yield with original timing
      if (this.entries.length === 0) return;

      const startTime = Date.now();
      const firstTimestamp = this.entries[0].timestamp;

      for (const entry of this.entries) {
        // Calculate delay based on original timestamp
        const elapsedMs = Date.now() - startTime;
        const targetMs = entry.timestamp - firstTimestamp;
        const delayMs = targetMs - elapsedMs;

        if (delayMs > 0) {
          await this.sleep(delayMs);
        }

        yield entry;
      }
    } else if (mode === "step") {
      // Manual step-by-step
      // For now, just yield with small delay
      for (const entry of this.entries) {
        yield entry;
        await this.sleep(100); // Small delay for step mode
      }
    }
  }

  /**
   * Get statistics about the recorded session.
   *
   * @returns PlayerStats
   */
  getStats(): PlayerStats {
    if (this.entries.length === 0) {
      return {
        count: 0,
        duration_ms: 0,
        first_timestamp: null,
        last_timestamp: null,
      };
    }

    const firstTimestamp = this.entries[0].timestamp;
    const lastTimestamp = this.entries[this.entries.length - 1].timestamp;
    const durationMs = lastTimestamp - firstTimestamp;

    return {
      count: this.entries.length,
      duration_ms: durationMs,
      first_timestamp: firstTimestamp,
      last_timestamp: lastTimestamp,
    };
  }

  /**
   * Get all entries (for inspection/debugging).
   *
   * @returns Array of ChunkLogEntry
   */
  getEntries(): ChunkLogEntry[] {
    return [...this.entries];
  }

  /**
   * Helper to sleep for specified milliseconds.
   *
   * @param ms - Milliseconds to sleep
   */
  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}
